File: vision/qr_detector.py
"""
二维码识别与任务码解析模块
- 二维码识别
- 任务码格式校验与解析 (XXX+XXX+XXX+XXX)
"""
import cv2


class QRDetector:
    """二维码识别器"""

    _BBOX_COLOR = (255, 0, 0)
    _BBOX_THICKNESS = 3
    _CENTER_COLOR = (0, 255, 0)
    _CENTER_RADIUS = 5

    def __init__(self):
        self.decoder = cv2.QRCodeDetector()

    def _decode(self, img):
        """统一解码接口，返回 (data, bbox)"""
        data, bbox, _ = self.decoder.detectAndDecode(img)
        return data, bbox

    def detect(self, img):
        """识别二维码，返回字符串；无二维码返回None"""
        if img is None:
            return None
        data, _ = self._decode(img)
        return data if data else None

    def detect_and_draw(self, img):
        """识别二维码并绘制标记"""
        if img is None:
            return None
        data, bbox = self._decode(img)
        if not data or bbox is None:
            return None

        n = len(bbox[0])
        for j in range(n):
            p1 = tuple(bbox[0][j])
            p2 = tuple(bbox[0][(j + 1) % n])
            cv2.line(img, p1, p2, self._BBOX_COLOR, self._BBOX_THICKNESS)

        xs = [p[0] for p in bbox[0]]
        ys = [p[1] for p in bbox[0]]
        cx = int(sum(xs) / n)
        cy = int(sum(ys) / n)
        cv2.circle(img, (cx, cy), self._CENTER_RADIUS, self._CENTER_COLOR, -1)
        return data


class TaskCodeParser:
    """
    任务码解析器
    比赛规则：四组三位数以"+"连接
    第1组: 第一批3个物料颜色和搬运顺序
    第2组: 第一批物料在粗加工区/暂存区的放置环号
    第3组: 第二批3个物料颜色和搬运顺序
    第4组: 第二批物料在粗加工区的放置环号
    例: 156+123+516+231
    """
    @staticmethod
    def parse(qr_data):
        """
        解析任务码

        :return: (batch1_colors, batch1_positions, batch2_colors, batch2_positions)
                 解析失败返回None
        """
        if not qr_data or '+' not in qr_data:
            return None
        try:
            parts = qr_data.split('+')
            if len(parts) != 4:
                return None
            # 每组必须3位数字
            for p in parts:
                if len(p) != 3 or not p.isdigit():
                    return None

            batch1_colors = [int(parts[0][0]), int(parts[0][1]), int(parts[0][2])]
            batch1_positions = [int(parts[1][0]), int(parts[1][1]), int(parts[1][2])]
            batch2_colors = [int(parts[2][0]), int(parts[2][1]), int(parts[2][2])]
            batch2_positions = [int(parts[3][0]), int(parts[3][1]), int(parts[3][2])]

            # 颜色编号和环号必须1~6
            for c in batch1_colors + batch2_colors:
                if c < 1 or c > 6:
                    return None
            for r in batch1_positions + batch2_positions:
                if r < 1 or r > 6:
                    return None

            return batch1_colors, batch1_positions, batch2_colors, batch2_positions
        except Exception:
            return None

File: vision_cpp/src/qr_detector.cpp
/**
 * @file qr_detector.cpp
 * @brief 二维码识别与任务码解析实现
 *
 * QRDetector 使用 OpenCV QRCodeDetector 进行检测与解码，
 * detect_and_draw 在图像上绘制蓝色边框和绿色中心点。
 * TaskCodeParser 将 "XXX+YYY+ZZZ+WWW" 格式字符串解析为 TaskCode。
 */
#include "qr_detector.hpp"

#include <cctype>

/** @brief 默认构造函数 */
QRDetector::QRDetector() = default;

/**
 * @brief 检测并解码二维码
 * @param img 输入图像（BGR）
 * @return 解码字符串；未检测到或图像为空返回 std::nullopt
 */
std::optional<std::string> QRDetector::detect(cv::Mat& img) 
{
    if (img.empty())
        return std::nullopt;

    cv::Mat points;
    /*5. 和你代码链路关联
    相机读到 frame → 丢进 detectAndDecode → 判断返回字符串非空 → 调用你注册的 lambda 回调，把二维码数据抛给上层 UI / 串口。*/
    std::string data = decoder_.detectAndDecode(img, points);

    if (data.empty())
        return std::nullopt;

    return data;
}

/**
 * @brief 检测并解码二维码，绘制边框和中心点
 * @param img 输入图像（BGR），将被修改
 * @return 解码字符串；未检测到返回 std::nullopt
 * @note 蓝色线段连接角点，绿色实心圆标注中心
 */
std::optional<std::string> QRDetector::detect_and_draw(cv::Mat& img) {
    if (img.empty())
        return std::nullopt;

    cv::Mat points;
    std::string data = decoder_.detectAndDecode(img, points);
    if (data.empty() || points.empty())
        return std::nullopt;

    // 绘制蓝色边框：连接相邻角点
    int n = static_cast<int>(points.total());
    for (int j = 0; j < n; j++) {
        cv::Point2f p1 = points.at<cv::Point2f>(j);
        cv::Point2f p2 = points.at<cv::Point2f>((j + 1) % n);

        cv::line(img,
                 cv::Point(static_cast<int>(p1.x), static_cast<int>(p1.y)),
                 cv::Point(static_cast<int>(p2.x), static_cast<int>(p2.y)),
                 cv::Scalar(255, 0, 0), 3);
    }

    // 计算并绘制绿色中心点
    float sum_x = 0.0f, sum_y = 0.0f;
    for (int i = 0; i < n; i++) {
        cv::Point2f p = points.at<cv::Point2f>(i);
        sum_x += p.x;
        sum_y += p.y;
    }
    int cx = static_cast<int>(sum_x / n);
    int cy = static_cast<int>(sum_y / n);
    cv::circle(img, cv::Point(cx, cy), 5, cv::Scalar(0, 255, 0), -1);

    return data;
}

namespace {

/**
 * @brief 3位数字字符串转 int 数组
 * @param s 长度为3的数字字符串
 * @return 3个整数的向量
 */
std::vector<int> extract_digits(const std::string& s) noexcept {
    return {s[0] - '0', s[1] - '0', s[2] - '0'};
}

/**
 * @brief 校验数字范围是否在 1~6
 * @param digits 数字向量
 * @return 全部合法返回 true
 */
bool is_valid_digit_range(const std::vector<int>& digits) noexcept {
    for (int d : digits) {
        if (d < 1 || d > 6) return false;
    }
    return true;
}

} // namespace

/**
 * @brief 解析二维码字符串为 TaskCode
 * @param qr_data 格式 "XXX+YYY+ZZZ+WWW"，4段3位数字
 * @return 解析结果；格式非法、非数字、数值越界返回 std::nullopt
 * @note 每位数字须在 1~6 范围内（颜色/位置编号）
 */
std::optional<TaskCode> TaskCodeParser::parse(const std::string& qr_data) {
    if (qr_data.empty())
        return std::nullopt;
    if (qr_data.find('+') == std::string::npos)
        return std::nullopt;

    // 按 '+' 分割为4段
    std::vector<std::string> parts;
    std::string current;
    for (char c : qr_data) {
        if (c == '+') {
            parts.push_back(current);
            current.clear();
        } else {
            current += c;
        }
    }
    parts.push_back(current);

    // 必须恰好4段
    if (parts.size() != 4)
        return std::nullopt;

    // 每段必须为3位数字
    for (const auto& p : parts) {
        if (p.length() != 3)
            return std::nullopt;
        for (char c : p) {
            if (!std::isdigit(static_cast<unsigned char>(c)))
                return std::nullopt;
        }
    }

    TaskCode code;
    code.batch1_colors = extract_digits(parts[0]);
    code.batch1_positions = extract_digits(parts[1]);
    code.batch2_colors = extract_digits(parts[2]);
    code.batch2_positions = extract_digits(parts[3]);

    // 校验数字范围 1~6
    if (!is_valid_digit_range(code.batch1_colors)) return std::nullopt;
    if (!is_valid_digit_range(code.batch2_colors)) return std::nullopt;
    if (!is_valid_digit_range(code.batch1_positions)) return std::nullopt;
    if (!is_valid_digit_range(code.batch2_positions)) return std::nullopt;

    return code;
}
